File: LeetCode/Math/Python/Leetcode795-number-of-subarrays-with-bounded-maximum.py
class Solution:
    def numSubarrayBoundedMax(self, nums: List[int], left: int, right: int) -> int:
        n = len(nums)
        def calcu(mx):
            nonlocal n
            res, i = 0,0
            while i < n:
                if nums[i] > mx:
                    i += 1
                    continue
                j = i + 1
                while j < n and nums[j] <= mx: j += 1
                dis = j - i
                res += (dis + 1) * dis // 2
                i = j + 1
            return res
        return calcu(right) - calcu(left - 1) 

File: LeetCode/Math/Rust/Leetcode795-number-of-subarrays-with-bounded-maximum.rs
impl Solution {
  pub fn num_subarray_bounded_max(nums: Vec<i32>, left: i32, right: i32) -> i32 {
      let n = nums.len();
      let calcu = |mx:i32| -> i32 {
          let (mut res, mut i) = (0, 0);
          while i < n {
              if nums[i] > mx {
                  i += 1;
                  continue;
              }
              let mut j = i + 1;
              while j < n && nums[j] <= mx {
                  j += 1;
              }
              let dis = j - i;
              res += (dis + 1) * dis / 2;
              i = j;
          }
          res as i32
      };
      calcu(right) - calcu(left - 1)
  }
}